cmd/nerdctl/run_linux.go

```diff
@@ -217,3 +229,85 @@ func withOOMScoreAdj(score int) oci.SpecOpts {
 		return nil
 	}
 }
+
+func generatePIDOpts(ctx context.Context, client *containerd.Client, pid string) ([]oci.SpecOpts, error) {
+	opts := make([]oci.SpecOpts, 0)
+	pid = strings.ToLower(pid)
+
+	switch pid {
+	case "":
+		// do nothing
+	case "host":
+		opts = append(opts, oci.WithHostNamespace(specs.PIDNamespace))
+		if rootlessutil.IsRootless() {
+			opts = append(opts, withBindMountHostProcfs)
+		}
+	default: // container:<id|name>
+		parsed := strings.Split(pid, ":")
+		if len(parsed) < 2 || parsed[0] != "container" {
+			return nil, fmt.Errorf("invalid pid namespace. Set --pid=[host|container:<name|id>")
+		}
+
+		containerName := parsed[1]
+		var targetCon containerd.Container
+		walker := &containerwalker.ContainerWalker{
+			Client: client,
+			OnFound: func(ctx context.Context, found containerwalker.Found) error {
+				if found.MatchCount > 1 {
+					return fmt.Errorf("multiple IDs found with provided prefix: %s", found.Req)
+				}
+				targetCon = found.Container
+				return nil
+			},
+		}
+		matchedCount, err := walker.Walk(ctx, containerName)
+		if err != nil {
+			return nil, err
+		}
+		if matchedCount < 1 {
+			return nil, fmt.Errorf("no such container: %s", containerName)
+		}
+
+		o, err := generateSharingPIDOpts(ctx, targetCon)
+		if err != nil {
+			return nil, err
+		}
+
+		opts = append(opts, o...)
+	}
+
+	return opts, nil
+}
+
+func generatePIDCOpts(ctx context.Context, client *containerd.Client, pid string) ([]containerd.NewContainerOpts, error) {
+	pid = strings.ToLower(pid)
+
+	cOpts := make([]containerd.NewContainerOpts, 0)
+	parsed := strings.Split(pid, ":")
+	if len(parsed) < 2 || parsed[0] != "container" {
+		// no need to save pid options
+		return cOpts, nil
+	}
+
+	containerName := parsed[1]
+	walker := &containerwalker.ContainerWalker{
+		Client: client,
+		OnFound: func(ctx context.Context, found containerwalker.Found) error {
+			if found.MatchCount > 1 {
+				return fmt.Errorf("multiple IDs found with provided prefix: %s", found.Req)
+			}
+			cOpts = append(cOpts, containerd.WithAdditionalContainerLabels(map[string]string{
+				labels.PIDContainer: containerName,
+			}))
```

Comment:
We already have `withInternalLabels`, should be put together

Reply:
When a new label is needed, the best way is to add a new parameter of `withInternalLabels()`?
And if so, `--pid` flags is processed in `setPlatformOptions()`. In order to pass into the `withInternalLabels()`, the number of `setPlatformOptions()`'s return value should be increased.

Reply:
> And if so, --pid flags is processed in setPlatformOptions().

I think the `--pid` flags should be platform independent, why not consider moving it out of `setPlatformOptions`?

Reply:
Previsouly, `--pid` only works on linux.
https://github.com/containerd/nerdctl/blob/c6517c7f9a178950b965748f6698f5ca4e76e318/cmd/nerdctl/run_linux.go#L135-L144

The reason, why it only works on the linux, is that it's too sensitive with namespace suppport.

https://github.com/containerd/nerdctl/blob/d15ada7bec3df3a4e2b6d78bdc9f5d518aaf9d44/cmd/nerdctl/run.go#L1184-L1187

```golang
// WithHostNamespace allows a task to run inside the host's linux namespace
func WithHostNamespace(ns specs.LinuxNamespaceType) SpecOpts {
	return func(_ context.Context, _ Client, _ *containers.Container, s *Spec) error {
		setLinux(s)
		for i, n := range s.Linux.Namespaces {
			if n.Type == ns {
				s.Linux.Namespaces = append(s.Linux.Namespaces[:i], s.Linux.Namespaces[i+1:]...)
				return nil
			}
		}
		return nil
	}
}
```

To avoid a compile error, I placed `reconfigPIDContainer()` in `cmd/nerdctl/start.go` but I think it caused confusing.
So, I added a condition to check runtime os now. https://github.com/containerd/nerdctl/blob/8680bdb9a24d0c99fba6682af7473124469cb2fd/cmd/nerdctl/start.go#L231

Anyway, in my opinion, `--pid` currently can work on linux only.
Because of it, `--pid` should be processed in `setPlatformOptions()`. But, I really agree with you. `setPlatformContainerOptions()` can make a fragmentation and cause to find a target container twice.
I tried to find a better way to improve it according to your review, but I couldn't. I'm really sorry.